### ISSUE TITLE
Visibility of a node should apply to all its children

### ISSUE THREAD

**Issue by 2plus2makes5** [godotengine/godot-proposals#2218]
### Describe the project you are working on
Dungeon RPG

### Describe the problem or limitation you are having in your project
If a node has a mix of Spatial, Node2D and Control children its visibility only applies to the children of the same type(i mean 3d, 2d or control).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The visibility of a node should apply to all its children.

Also consider adding visibility to "Node", 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
The visibility of a node applies to all its children.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Well one can toggle the visibility of each child by code, but it shouldn't the default way of doing things in my opinion.

### Is there a reason why this should be core and not an add-on in the asset library?
It's core i guess.

**Comment by Calinou** [comment]
I think this is done by design. If we were to change this, I wouldn't be surprised if some other people complain about it as they rely on the current behavior :slightly_smiling_face:

**Comment by KoBeWi** [comment]
Visibility does work between Nodes2D and Controls.

2D and 3D are separate, so current behavior makes sense.

**Comment by 2plus2makes5** [comment]
I imagined that it was by design, but i my opinion it's more natural that all the children follow the same behaviour and eventually make individual changes in code instead of the opposite like now.
For me all the children of a node are part of the same entity so any change to that entity should be applied to all the children.

Let's make an example, if a 3d enemy has 2d stats on screen, i'm pretty sure that most of the people who want the 3d enemy to disappear expect the 2d stuff to disappear too.

**Comment by KoBeWi** [comment]
Yeah but usually you wrap the 2D stats in a 3D node to show it above enemy (because 2D nodes show in the screen space). Or if you have e.g. jRPG battle system, UI is not part of enemy nodes at all. And if you remove the enemy (free it), the child 2D nodes would disappear too.

**Comment by 2plus2makes5** [comment]
@KoBeWi 
3D games can have 2d interface too and i was thinking of disappearing enemies, not killed one, but ok let's scratch that, it was just an hypothetical case.

Let's talk about a real case, I have a scene manager and each phase of the game(exploration, fight, menu etc) has its own scene and i did my best to make each scene as self contained and indipendent from the others as possible.

Most of the scenes have all kind of elements, when for example i pass from exploration to fight i obviously want to keep the data of the exploration(map, enemies, items etc) so i just deactivate the scene, make it invisible and pass to fight, but making the root invisible isn't enough and i have to make many elements invisible by code, otherwise some elements remain on screen, and that applies to all the scenes that have different kinds of nodes.

**Comment by pycbouh** [comment]
You can temporary remove your scene from the tree, since it doesn't need to be updated in that phase.

**Comment by 2plus2makes5** [comment]
@pycbouh 
Yeah i'm considering that solution, it would also be better for performance right?

The point remains though, in my opinion we should be able to mix different nodes or not, a middle ground that let us mix nodes but with catches is not elegant at least.

That's me though.

**Comment by pycbouh** [comment]
> The point remains though, in my opinion we should be able to mix different nodes or not, a middle ground that let us mix nodes but with catches is not elegant at least.

Think about the opposite situation, where we do assume this "smart" behavior. The workaround would be far less elegant for those cases. Currently all you need to do is to be explicit about what you want to have hidden. In the opposite situation you'd have to "unhide" nodes wrongly affected by it, which is a much uglier code pattern, as you actively fight against a system.

Current situation doesn't assume user's intention which allows everyone to be explicit and expressive in their intention, even though it may lead to more verbose code in some system designs. But in the opposite case we'd have to assume that the user has intended what you believe the user most likely intends, thus fixing everyone to the same logic pattern. I would prefer for the engine to not make any such assumptions.

Also keep in mind that we have another cases of cross-context behavior that people might "expect" to work as well: regular nodes don't carry any relative and visual properties across them. This means that grouping visual nodes by a regular node is equaly impossible.

At best, it can be an optional (and off by default) property of `Node` to carry specific properties across contexts. But even that can be too much, as you can work around this by designing your systems with current behavior in mind instead.

**Comment by 2plus2makes5** [comment]
I guess it's a matter of perspective, for you it's wrong that the visibility of a node affets all its children and you think that unhiding individual children is fighting against the system, for me it's wrong that the visibility of a node doesn't affect all its children and to me being forced to specify the same visibility as the parent is fighting against the system.

If a node already does what we want we don't add code, we add code if we want it to do something different, so to me it's more natural to code only if we want children to have a different behaviour, not to have the same behaviour of the parent node.

**Comment by dalexeev** [comment]
Just a clarification: besides `CanvasItem` (`Control` and `Node2D`) and `Node3D`, the `visible` property is separately defined for `Window`.

**Comment by arnaucmd** [comment]
Somebody knows if there is a simple solution to this problem? For now I can't find a way to change the visibility of all the children by just toggling the visibility of the parent. I agree it should be the default behaviour or at least just have the option to set the parent to affect all children visibility with an option or similar.
